kaspresso/src/main/kotlin/com/kaspersky/kaspresso/interceptors/behavior/impl/autoscroll/FallbackAutoScrollToAction.kt

```diff
@@ -0,0 +1,91 @@
+package com.kaspersky.kaspresso.interceptors.behavior.impl.autoscroll
+
+import android.util.Log
+import android.view.View
+import android.view.ViewParent
+import android.widget.HorizontalScrollView
+import android.widget.ListView
+import android.widget.ScrollView
+import androidx.core.widget.NestedScrollView
+import androidx.test.espresso.PerformException
+import androidx.test.espresso.UiController
+import androidx.test.espresso.ViewAction
+import androidx.test.espresso.matcher.ViewMatchers
+import androidx.test.espresso.util.HumanReadables
+import org.hamcrest.Matcher
+import org.hamcrest.Matchers
+
+class FallbackAutoScrollToAction : ViewAction {
+    override fun getConstraints(): Matcher<View> {
+        return Matchers.allOf(
+            ViewMatchers.withEffectiveVisibility(ViewMatchers.Visibility.VISIBLE),
+            ViewMatchers.isDescendantOfA(
+                Matchers.anyOf(
+                    ViewMatchers.isAssignableFrom(NestedScrollView::class.java),
+                    ViewMatchers.isAssignableFrom(ScrollView::class.java),
+                    ViewMatchers.isAssignableFrom(HorizontalScrollView::class.java),
+                    ViewMatchers.isAssignableFrom(ListView::class.java)
+                )
+            )
+        )
+    }
+
+    private fun View?.isScrollable(): Boolean =
+        (this is ScrollView || this is NestedScrollView || this is HorizontalScrollView || this is ListView)
+
+    private tailrec fun View?.findFirstParentScrollableView(lastView: View): View? {
+        return if (this == lastView) {
+            if (this.isScrollable()) this else null
+        } else {
+            if (this.isScrollable()){
+                this
+            } else {
+                this?.parent?.findFirstView()?.findFirstParentScrollableView(lastView)
+            }
+        }
+    }
+
+    private tailrec fun ViewParent?.findFirstView(): View? {
+        return if (this is View) {
+            this
+        } else {
+            this?.parent.findFirstView()
+        }
+    }
+
+    override fun perform(uiController: UiController, view: View) {
+        if (ViewMatchers.isDisplayingAtLeast(100).matches(view)) {
+            Log.i(TAG, "View is already displayed. Returning.")
+            return
+        }
+
+        val scrollView = view.findFirstParentScrollableView(view.rootView)
+
+        when (scrollView is HorizontalScrollView) {
+            true -> scrollView.scrollTo(view.right + scrollView.paddingEnd, 0)
```

Comment:
I am wondering about the opposite case: What if we want to scroll UP/LEFT because the initial position of the scroll is in the middle and the element we want to scroll to is not DOWN/RIGHT?

Have some idea to implement this... need to test it

Reply:
@Vacxe what do you think?

Reply:
1) Will it work for left-right or bottom-top scroll?
2) Will it work for LTR orientation?
I think we need abstractions here or just we can make this view specific: RightLeftAutoscrollAction, TopBottomAutoscrollAction. What do you think?

Reply:
@sergio-sastre did you test it on LTR orientation?)

Reply:
@sergio-sastre Are you going to implement cases when scrollView is not a HorizontalScrollView?

Reply:
@eakurnikov the LRT is a very good point. I need to verify it works.
The other cases (e.g. scrollView is not a HorizontalScrollView) are covered in the false-clause

Reply:
- [ ] Use abstractions for RTLAutoscrollAction, LTRAutoscrollAction

Reply:
Hi there!
Since I updated to 1.4.1, it seems I am not able to run the app itself. That is very inconvenient, since I cannot run anything, including the tests, to verify if changes in the code are correct...
<img width="700" alt="Screenshot 2022-03-17 at 07 58 36" src="https://user-images.githubusercontent.com/6097181/158754110-441c1f5e-c4be-4396-a6f9-22ee6ea67e70.png">

Any help on this?

Reply:
Hi there! Sorry for not finding the time for this, I've been very busy lately. I'll resume my work on the opened tickets **on 16th May**.
I hope it'll help make Kaspresso even greater :)

Reply:
@eakurnikov 
I tested the LTR. It works! Here the steps and the video
1) I enabled pseudolocales
2) I set Locale to ar_XB 
3) I executed the ScrollViewWithPaddingPassingTests
And here the videos

RTL Overview  (no test running)-> HorizontalScrollView starting from Rigth

https://user-images.githubusercontent.com/6097181/170633182-a7be22f5-2e64-4a03-b70f-591cd9bd48b2.mov

RTL -> running test ScrollViewWithPaddingPassingTests#click_last_button_in_HorizontalScrollView_and_then_first_one()

https://user-images.githubusercontent.com/6097181/170633420-1516b197-2182-423c-92a0-741a34327b43.mov

Reply:
> I am wondering about the opposite case: What if we want to scroll UP/LEFT because the initial position of the scroll is in the middle and the element we want to scroll to is not DOWN/RIGHT?
> 
> Have some idea to implement this... need to test it

- [ ] Create test that scrolls to the middle, and wants to click on the first element

Reply:
@sergio-sastre sorry, is there any problem now with running the app? I've just check and it works for me.